kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/Dek.java

```diff
@@ -0,0 +1,447 @@
+ * instances are not. {@link Encryptor Encryptor}s and {@link Decryptor Decryptor}s can be allocated from a common {@link Dek} instance that's shared between multiple threads,
+ * but once allocated those cryptors should remain localised to a single thread for the duration of their lifetime.</p>
+ *
+ * <h2 id="destruction">DEK destruction</h2>
+ * <p>A {@code Dek} can be destroyed, which will destroy
+ * the underlying key material if possible.</p>
+ * <p>To do this both {@link #destroyForEncrypt()} and {@link #destroyForDecrypt()}
+ * need to be called on the key instance.
+ * Even once both calls have been made the key is not destroyed immediately.
+ * Instead calls to {@link #encryptor(int)} and {@link #decryptor()} will start to throw {@link DestroyedDekException},
+ * but existing {@link Encryptor Encryptor} and {@link Decryptor Decryptor} instances will be able to continue.
+ * The key will only be destroyed once the {@code close()} method has been called on all existing (en|de)cryptors.</p>
+ *
+ * @param <E> The type of encrypted DEK.
+ */
+@ThreadSafe
+public final class Dek<E> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Dek.class);
+
+    private static final Map<Class<? extends Destroyable>, Boolean> LOGGED_DESTROY_FAILED = new ConcurrentHashMap<>();
+
+    private final E edek;
+    // Note: checks for outstandingCryptors==END <em>happens-before</em> changes to atomicKey.
+    // It's this that provides a guarantee that a cryptor should never see a
+    // destroyed or null key, but that keys should be destroyed and nullieifed as soon
+    // as possible once all outstanding cryptors are closed.
+    private final AtomicReference<SecretKey> atomicKey;
+
+    private final AtomicLong remainingEncryptions;
+
+    // 1 start
+    // +1=2 encryptor
+    // +1=3 encryptor
+    // +1=4 encryptor
+    // -1=3 encryptor.finish // return
+    // negate=-3 destroy()// updateAndGet(if current value < 0 don't change, otherwise negate)
+    // +1=-2 encryptor.finish
+    // +1=-1 encryptor.finish => key.destroy
+    final AtomicLong outstandingCryptors;
+
+    public static final long START = combine(1, 1);
+    public static final long END = combine(-1, -1);
+    private final CipherSpec cipherSpec;
+
+    static long combine(int encryptors, int decryptors) {
+        return ((long) encryptors) << Integer.SIZE | 0xFFFFFFFFL & decryptors;
+    }
+
+    static int encryptors(long combined) {
+        return (int) (combined >> Integer.SIZE);
+    }
+
+    static int decryptors(long combined) {
+        return (int) combined;
+    }
+
+    static long acquireEncryptor(long combined) {
+        int encryptors = encryptors(combined);
+        int decryptors = decryptors(combined);
+        if (encryptors > 0) {
+            return combine(encryptors + 1, decryptors);
+        }
+        else {
+            return combine(encryptors, decryptors);
```

Comment:
should this be encryptors - 1 like we do with the decryptors in `acquireDecryptor`?

Reply:
Oh I guess it's to prevent handing out encryptors post encryptor destroy

Reply:
Thanks! This made me look again and find a bug. I've documented the reference counting protocol a little better now too; hopefully it helps. Let me know if you can see any ways of making this reference counting easier to understand.